michbl/assignment2/BinaryTree.hpp

```diff
@@ -84,111 +82,99 @@ class BinaryTree {
+    bool printAncestors(const Node<T> *node, const T &key) {
         if (node == nullptr) {
-            return;
+            return false;
         }
         if (node->key == key) {
-            found = true;
-            return;
-        }
-        if (!found) {
-            printAncestors(node->left, key, found);
-        }
-        if (!found) {
-            printAncestors(node->right, key, found);
+            return true;
         }
-        if (found) {
+
+        if (printAncestors(node->left, key) || printAncestors(node->right, key)) {
             if (node == this->root) {
-                cout << node->key << endl;
+                cout << node->key;
             }
             else {
                 cout << node->key << ", ";
             }
         }
     }
 
-    bool bothKeysAreInSubtree(Node<T> *node, T key1, T key2) {
-        bool found1 = false, found2 = false;
-        bothKeysAreInSubtreeAux(node->left, key1, key2, found1, found2);
-        bothKeysAreInSubtreeAux(node->right, key1, key2, found1, found2);
-        return found1 && found2;
+    bool isAncestor(const Node<T> *node, const T &key1, const T &key2) {
+        bool found1 = keyIsInSubtree(node->left, key1) || keyIsInSubtree(node->right, key1);
+        bool found2 = keyIsInSubtree(node->left, key2) || keyIsInSubtree(node->right, key2);
+        return  found1 && found2;
     }
 
-    void bothKeysAreInSubtreeAux(Node<T> *node, T key1, T key2, bool &found1, bool &found2) {
+    bool keyIsInSubtree(const Node<T> *node, T key) {
         if (node == nullptr) {
-            return;
-        }
-        if (node->key == key1) {
-            found1 = true;
+            return false;
         }
-        if (node->key == key2) {
-            found2 = true;
+        if (node->key == key) {
+            return true;
         }
-        bothKeysAreInSubtreeAux(node->left, key1, key2, found1, found2);
-        bothKeysAreInSubtreeAux(node->right, key1, key2, found1, found2);
+        return keyIsInSubtree(node->left, key) || keyIsInSubtree(node->right, key);
     }
 
-    void printLowestCommonAncestor(Node<T> *node, T key1, T key2, bool &found) {
+    // The return value denotes weather the LCA was already found or not
+    bool printLowestCommonAncestor(const Node<T> *node, const T &key1, const T &key2) {
         if (node == nullptr) {
-            return;
-        }
-        if (!found) {
-            printLowestCommonAncestor(node->left, key1, key2, found);
-        }
-        if (!found) {
-            printLowestCommonAncestor(node->right, key1, key2, found);
-        }
-        if (!found && bothKeysAreInSubtree(node, key1, key2)) {
-            found = true;
-            cout << node->key << endl;
+            return false;
+        }
+        bool foundInLeft = printLowestCommonAncestor(node->left, key1, key2);
+        bool foundInRight = printLowestCommonAncestor(node->right, key1, key2);
+        if (!foundInLeft && !foundInRight) {

```

Comment:
The check `!foundInLeft && !foundInRight` is exactly the opposite of `foundInLeft || foundInRight` (according to De Morgans law). So this is a (minor) code duplication!
One option is to put it inside an else:
```C++
if (!foundInLeft && !foundInRight) {
    ...
} else {
    return true;
}
```
Or even remove the else because the first block always returns:
```C++
if (!foundInLeft && !foundInRight) {
    ...
}
return true;
```

But my suggestion is to flip the cases and return early. This way you will also have less nesting, and less need to navigate between all the ifs and elses:
```C++
if (foundInLeft || foundInRight) {
    // Both keys found in one of the subtrees, meaning this is an ancestor bot not
    // the lowest common ancestor.
    return true;
}
if (isAncestor(node, key1, key1)) {
    // This is the lowest common ancestor.
    cout << node->key;
    return true;
}
// This is not an ancestor at all.
return false;
```
This (optional) style avoids nesting and even without the comments, it is read as a list of cases and not a forest of if/elses (and no duplication). It has its downsides (e.g. many return points) but it is an option to consider.